lib/msf/core/payload/android.rb

```diff
@@ -31,25 +36,53 @@ def java_string(str)
     [str.length].pack("N") + str
   end
 
-  def apply_options(classes)
+  def apply_options(classes, opts)
     timeouts = [
       datastore['SessionExpirationTimeout'].to_s,
       datastore['SessionCommunicationTimeout'].to_s,
       datastore['SessionRetryTotal'].to_s,
       datastore['SessionRetryWait'].to_s
     ].join('-')
-    string_sub(classes, 'TTTT                                ', 'TTTT' + timeouts)
+    if opts[:stageless]
+      config = generate_config_hex(opts)
+      string_sub(classes, 'UUUU' + ' ' * 8191, 'UUUU' + config)
+    end
+    if opts[:ssl]
+      verify_cert_hash = get_ssl_cert_hash(datastore['StagerVerifySSLCert'],
+                                           datastore['HandlerSSLCert'])
+      if verify_cert_hash
+        hash = 'WWWW' + verify_cert_hash.unpack("H*").first
+        string_sub(classes, 'WWWW                                        ', hash)
+      end
+    end
+    string_sub(classes, 'ZZZZ' + ' ' * 512, 'ZZZZ' + payload_uri)
+    string_sub(classes, 'TTTT' + ' ' * 48, 'TTTT' + timeouts)
+  end
+
+  def generate_config_hex(opts={})
+    opts[:uuid] ||= generate_payload_uuid
+
+    config_opts = {
+      ascii_str:  true,
+      arch:       opts[:uuid].arch,
+      expiration: datastore['SessionExpirationTimeout'].to_i,
+      uuid:       opts[:uuid],
+      transports: [transport_config(opts)]
+    }
+
+    config = Rex::Payloads::Meterpreter::Config.new(config_opts)
+    config.to_b.unpack('H*').first
   end
 
   def string_sub(data, placeholder="", input="")
     data.gsub!(placeholder, input + ' ' * (placeholder.length - input.length))
   end
 
-  def generate_cert
+  def sign_jar(jar)
     x509_name = OpenSSL::X509::Name.parse(
-      "C=Unknown/ST=Unknown/L=Unknown/O=Unknown/OU=Unknown/CN=Unknown"
-      )
-    key  = OpenSSL::PKey::RSA.new(1024)
+      "C=US/O=Android/CN=Android Debug"
+    )
+    key  = OpenSSL::PKey::RSA.new(2048)
```

Comment:
Hey @timwr any context on the change from 1024 to 2048 here? Just noticed that it's considerably slower than 1024:

```
Time for 1024 PKey generation: 0.005687
Time for 2048 PKey generation: 0.150518
```

That's not a big issue, but on boot up we seem to generate a few android jars - so it all adds up to about a second of overhead. Any insight would be appreciated :+1:

Reply:
Low certificate strength could be used as an indicator of compromised packaging, `2048` set in 2016 seems like an appropriate min key and soon even larger may be appropriate.

The cert being created on startup is a separate issue, possibly reducing how ofter a new cert is needed could be considered or making it a non-blocking task would likely be a valuable compromise.

Reply:
@adfoster-r7 @jmartin-r7 I think I was just attempting to match the output of jarsigner. I don't think it makes much difference, and we can use jarsigner (and now apksigner) to resign the apk anyway. I suggest we reduce it to 1024 to increase the speed. I wonder if we could reduce it further by not generating the jar at all on startup.

Reply:
Thanks for the extra insight - we'll have a PR up that hopefully makes this a bit faster 🤞 
I'll cross-reference that PR here for visibility too

Reply:
Referencing - https://github.com/rapid7/metasploit-framework/pull/13148